### PR TITLE
Restore chunked download support.

### PR DESCRIPTION
**Pull request by craigcitro**
We lost support for chunked downloads in a merge; this PR restores chunked
download support, and adds tests.

In particular, since gsutil does its own chunking, we do one-shot downloads in
the case that `download.chunksize == 0`.

PTAL @thobrla -- i'm sure there are some corner cases I'm not yet thinking of.

@charlesccychen I think this should get you fixed up.

### CODE REVIEW

File: apitools/base/py/transfer.py [hunk 1]
```diff
@@ -344,7 +344,9 @@ def __GetChunk(self, start, end=None, additional_headers=None):
         self.EnsureInitialized()
         end_byte = end
         if self.total_size and end:
-            end_byte = min(end, self.total_size)
+            end_byte = min(end_byte, self.total_size)
+        # if self.chunksize > 0:
```
Comment (thobrla) [inline apitools/base/py/transfer.py]: I assume this is meant to be uncommented.

Comment (thobrla) [inline apitools/base/py/transfer.py]: In the _Transfer base class, we accept a chunksize kwarg and use it like so:
self.chunksize = chunksize or 1048576

Which means a chunksize of 0 as a kwarg won't be respected.

File: apitools/base/py/transfer_test.py [hunk 2]
```diff
@@ -12,6 +14,79 @@
...
+            self.assertEqual(base_url, make_request.call_args[0][1].url)
+            download_stream.seek(0)
+            self.assertEqual(string.ascii_lowercase * 2,
+                             download_stream.getvalue())
+
+    def testChunkedDownload(self):
+        bytes_http = object()
+        http = object()
+        download_stream = six.StringIO()
+        download = transfer.Download.FromStream(
```
Comment (thobrla) [inline apitools/base/py/transfer_test.py]: This download setup looks identical to the non-chunked one, so why would we expect different behavior from the http request?

File: apitools/base/py/transfer.py [hunk 3]
```diff
@@ -344,7 +344,9 @@ def __GetChunk(self, start, end=None, additional_headers=None):
...
+        #     end_byte = min(end_byte, start + self.chunksize)
```
Comment (thobrla) [inline apitools/base/py/transfer.py]: I realize that this may also have implications for GetRange, but I like Upload's explicitly named parameter better than a sentinel value of 0 chunksize, precisely to avoid the kinds of issues I mentioned above.

Don't feel strongly about it, though.

File: apitools/base/py/transfer.py (hunk 1 above)
Comment (craigcitro) [inline apitools/base/py/transfer.py]: yargh, good call.

File: apitools/base/py/transfer.py (hunk 3 above)
Comment (craigcitro) [inline apitools/base/py/transfer.py]: cleaned up.

File: apitools/base/py/transfer_test.py (hunk 2 above)
Comment (craigcitro) [inline apitools/base/py/transfer_test.py]: mocking at 1AM != good idea. :grinning:

**Comment by craigcitro** [comment]
PTAL

File: apitools/base/py/transfer.py [hunk 4]
```diff
@@ -420,7 +432,29 @@ def GetRange(self, start, end=None, additional_headers=None):
...
     def StreamInChunks(self, callback=None, finish_callback=None,
                        additional_headers=None):
-        """Stream the entire download."""
+        """Stream the entire download in chunks."""
+        self.StreamMedia(callback=callback, finish_callback=finish_callback,
+                         additional_headers=additional_headers,
+                         ignore_chunksize=False)
+
+    def StreamMedia(self, callback=None, finish_callback=None,
+                    additional_headers=None, ignore_chunksize=False):
```
Comment (thobrla) [inline apitools/base/py/transfer.py]: Here and elsewhere, use_chunks = true for consistency with Upload?

File: apitools/base/py/transfer.py [hunk 5]
```diff
@@ -378,7 +386,8 @@ def __ProcessResponse(self, response):
             self.stream.write('')
         return response
 
-    def GetRange(self, start, end=None, additional_headers=None):
+    def GetRange(self, start, end=None, additional_headers=None,
+                 ignore_chunksize=False):
```
Comment (thobrla) [inline apitools/base/py/transfer.py]: This will require a gsutil revision (and thus an apitools release), which is fine.

File: apitools/base/py/transfer.py [hunk 6]
```diff
@@ -339,12 +341,18 @@ def __SetRangeHeader(self, request, start, end=None):
...
 
-    def __GetChunk(self, start, end=None, additional_headers=None):
+    def __GetChunk(self, start, end=None, additional_headers=None,
+                   ignore_chunksize=False):
         """Retrieve a chunk, and return the full response."""
         self.EnsureInitialized()
         end_byte = end
-        if self.total_size and end:
-            end_byte = min(end, self.total_size)
+        if not ignore_chunksize:
```
Comment (thobrla) [inline apitools/base/py/transfer.py]: It feels a bit strange to have a function named GetChunk care about chunking.  Why not push the logic around "end" into a helper function that's called by StreamMedia + GetRange?

File: apitools/base/py/transfer.py [hunk 7]
```diff
@@ -420,7 +432,29 @@ def GetRange(self, start, end=None, additional_headers=None):
...
+
+        Args:
+          callback: (default: None) Callback to call as each chunk is
+              completed.
+          finish_callback: (default: None) Callback to call when the
+              download is complete.
+          additional_headers: (default: None) Additional headers to
+              include in fetching bytes.
+          ignore_chunksize: (bool, default: False) If True, ignore
+              self.chunksize and attempt to stream this download in
```
Comment (thobrla) [inline apitools/base/py/transfer.py]: This is at odds with the function docstring.

File: apitools/base/py/transfer.py (hunk 4 above)
Comment (craigcitro) [inline apitools/base/py/transfer.py]: hah ... done.

File: apitools/base/py/transfer.py (hunk 5 above)
Comment (craigcitro) [inline apitools/base/py/transfer.py]: yeah -- it was really an accidental change in the default apitools behavior.

File: apitools/base/py/transfer.py (hunk 6 above)
Comment (craigcitro) [inline apitools/base/py/transfer.py]: my thought was that i didn't want anyone to accidentally skip this logic (say if there were another method added).

i don't mind the naming bit -- it's figuring out *what* chunk to get. :grin: but i'm pulling it out to add tests.

**Comment by charlesccychen** [comment]
LGTM

File: apitools/base/py/transfer.py (hunk 7 above)
Comment (craigcitro) [inline apitools/base/py/transfer.py]: good call.

**Comment by craigcitro** [comment]
PTAL

File: apitools/base/py/transfer.py [hunk 8]
```diff
@@ -339,14 +341,39 @@ def __SetRangeHeader(self, request, start, end=None):
         else:
             request.headers['range'] = 'bytes=%d-%d' % (start, end)
 
-    def __GetChunk(self, start, end=None, additional_headers=None):
+    def __ComputeEndByte(self, start, end=None, use_chunks=True):
+        """Compute the last byte to fetch for this request.
+
+        Note that this is potentially confusing in several ways:
+          * the value for the last byte is 0-based, eg "fetch 10 bytes
```
Comment (thobrla) [inline apitools/base/py/transfer.py]: Nit: this is per the HTTP spec for Range/Content-Range, may be useful to relate that.

File: apitools/base/py/transfer.py [hunk 9]
```diff
@@ -298,6 +298,8 @@ def InitializeDownload(self, http_request, http=None, client=None):
             http_request.url = client.FinalizeTransferUrl(http_request.url)
         url = http_request.url
         if self.auto_transfer:
+            end_byte = min(self.total_size or 0, self.chunksize) - 1
```
Comment (thobrla) [inline apitools/base/py/transfer.py]: Why not compute this with __ComputeEndByte?

File: apitools/base/py/transfer.py (hunk 8 above)
Comment (craigcitro) [inline apitools/base/py/transfer.py]: done.

File: apitools/base/py/transfer.py (hunk 9 above)
Comment (craigcitro) [inline apitools/base/py/transfer.py]: i don't believe in code reuse.

KIDDING! i forgot. done.

File: apitools/base/py/transfer.py [hunk 10]
```diff
@@ -339,14 +341,39 @@ def __SetRangeHeader(self, request, start, end=None):
...
+        Returns:
+          Last byte to use in a Range header, or None.
+        """
+        end_byte = end
+        if use_chunks:
+            alternate = start + self.chunksize - 1
+            end_byte = min(end_byte, alternate) if end_byte else alternate
+        if self.total_size:
+            alternate = self.total_size - 1
+            end_byte = min(end_byte, alternate) if end_byte else alternate
```
Comment (thobrla) [inline apitools/base/py/transfer.py]: I don't think this is harmful, but it is a semantic change where __GetChunk used to set a range header of:
bytes=0-

and now sets
bytes=0-22   # example assuming self.total_size == 23

in the case where use_chunks=False and bool(self.total_size)

File: apitools/base/py/transfer.py [hunk 11]
```diff
@@ -339,14 +341,39 @@ def __SetRangeHeader(self, request, start, end=None):
...
+
+        See the tests for more examples.
+
+        Args:
+          start: byte to start at.
+          end: (int or None, default: None) Suggested last byte.
+          use_chunks: (bool, default: True) If False, ignore self.chunksize.
+
+        Returns:
+          Last byte to use in a Range header, or None.
```
Comment (thobrla) [inline apitools/base/py/transfer.py]: This has some potentially strange interaction with negative start ranges.  Is it intended for use in that case?

File: apitools/base/py/transfer.py (hunk 10 above)
Comment (craigcitro) [inline apitools/base/py/transfer.py]: yeah, i thought about that when i was writing the tests; i also can't think of a case where it's harmful. 

well, actually, that's not true: if you had *mis-specified* the `total_size`, you'll now get an error trying to download instead of silently getting too many bytes. it hink that's better?

File: apitools/base/py/transfer.py (hunk 11 above)
Comment (craigcitro) [inline apitools/base/py/transfer.py]: so at least at this layer, the HTTP spec doesn't support negative start indices in the range.

in the case of something like `GetRange(-10)`, we normalize first (see `__NormalizeStartEnd`).

**Comment by craigcitro** [comment]
and i've confirmed that `gsutil test -p 100` passes with this code patched in.

File: apitools/base/py/transfer_test.py [hunk 12]
```diff
@@ -12,6 +14,146 @@
...
+               'response_range ({1})').format(
+                   request_range, response_range)
+        self.assertTrue(response_range.startswith(request_range), msg=msg)
+
+    def testComputeEndByte(self):
+        total_size = 100
+        chunksize = 10
+        download = transfer.Download.FromStream(
+            six.StringIO(), chunksize=chunksize, total_size=total_size)
+        self.assertEqual(9, download._Download__ComputeEndByte(0, end=50))
```
Comment (thobrla) [inline apitools/base/py/transfer_test.py]: Nit: chunksize - 1

Comment (craigcitro) [inline apitools/base/py/transfer_test.py]: touche!

File: apitools/base/py/transfer_test.py [hunk 13]
```diff
@@ -12,6 +14,146 @@
...
+                    content=string.ascii_lowercase,
+                    request_url='https://part.one/',
+                )
+            elif url == 'https://part.two/':
+                return http_wrapper.Response(
+                    info={
+                        'content-range': 'bytes 26-51/52',
+                        'status': http_client.OK,
+                    },
+                    content=string.ascii_lowercase,
```
Comment (thobrla) [inline apitools/base/py/transfer_test.py]: It would probably be good to differentiate the bytes in this case (so that the second response is different in content than the first)

Comment (craigcitro) [inline apitools/base/py/transfer_test.py]: good call, done.

**Comment by thobrla** [comment]
LGTM
